plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/purity/PurityChecker.kt

```diff
@@ -0,0 +1,180 @@
+            registry[key] = PurityStatus.IMPURE
+            dbg { "Purity check failed: $debugName is impure (all dependencies resolved)" }
+            resolveFunctionalDependencies(userFunc.name!!)
+        }
+        return result
+    }
+
+    private fun isExpressionPure(exp: ExpEmbedding): Boolean = when (exp) {
+        is PureExpEmbedding -> true
+
+        is Block -> exp.exps.all(::isExpressionPure)
+
+        is Assign -> if (exp.lhs is PlaceholderVariableEmbedding)
+            isExpressionPure(exp.rhs)
+        else
+            isExpressionPure(exp.lhs) && isExpressionPure(exp.rhs)
+
+        is PassthroughExpEmbedding -> isExpressionPure(exp.inner)
+
+        // Operators
+        is BinaryDirectResultExpEmbedding ->
+            isExpressionPure(exp.left) && isExpressionPure(exp.right)
+
+        is UnaryDirectResultExpEmbedding ->
+            isExpressionPure(exp.inner)
+
+        // Function calls
+        is MethodCall -> { // we are still missing self recursion
+            val isSelfCall = (exp.method.name == currentEvaluatedFunction?.name)
+
+            if (isSelfCall) {
+                exp.args.all(::isExpressionPure)
+            } else {
+                val mangledName = exp.method.name.mangled
+                val argsPure = exp.args.all(::isExpressionPure)
+                val operatorPure =
+                    PureOperatorWhitelist.isWhitelisted(mangledName) && argsPure
+                if (operatorPure) true
+                else {
+                    val callee = registry[exp.method.name]
+                    when (callee) {
+                        null -> {
+                            registry[exp.method.name] = PurityStatus.UNKNOWN
+                            addFunctionalDependency(exp.method.name)
+                            registry[currentEvaluatedFunction!!.name!!] = PurityStatus.CHECKING
+                            false
+                        }
+                        PurityStatus.CHECKING -> false
+                        PurityStatus.PURE -> true
+                        PurityStatus.IMPURE -> false
+                        else -> false
+                    }
+                }
+            }
+        }
+
+        // Single-expression If - just allow this for now
+        is If -> isExpressionPure(exp.condition) &&
+                isExpressionPure(exp.thenBranch) &&
+                isExpressionPure(exp.elseBranch)
+
+        is Goto -> true          // auto-generated return
+        is InhaleDirect -> true  // auto-generated check
+
+        // Local Variable Declarations
+        is Declare -> false
+        else -> false
+    }
+
+    private fun addFunctionalDependency(callee: MangledName) {
+        waiting.getOrPut(callee) { mutableSetOf() }
+            .add(currentEvaluatedFunction!!)
+
+        dbg { "${currentEvaluatedFunction!!.name!!.mangled} is now waiting on ${callee.mangled} (dependency not yet resolved)" }
+    }
```

Comment:
Except for inline functions, it generally doesn't make sense to check purity transitively: we only want to raise the error "impure function annotated pure" once.